lib/SIL/SILDeclRef.cpp

```diff
@@ -836,3 +836,45 @@ SILLocation SILDeclRef::getAsRegularLocation() const {
     return RegularLocation(getDecl());
   return RegularLocation(getAbstractClosureExpr());
 }
+
+SubclassScope SILDeclRef::getSubclassScope() const {
+  if (!hasDecl())
+    return SubclassScope::NotApplicable;
+
+  // If this declaration is a function which goes into a vtable, then it's
+  // symbol must be as visible as its class. Derived classes even have to put
+  // all less visible methods of the base class into their vtables.
+
+  auto *FD = dyn_cast<AbstractFunctionDecl>(getDecl());
+  if (!FD)
+    return SubclassScope::NotApplicable;
+
+  DeclContext *context = FD->getDeclContext();
+
+  // Methods from extensions don't go into vtables (yet).
+  if (context->isExtensionContext())
+    return SubclassScope::NotApplicable;
+
+  auto *classType = context->getAsClassOrClassExtensionContext();
+  if (!classType || classType->isFinal())
+    return SubclassScope::NotApplicable;
+
+  if (FD->isFinal() && !FD->getOverriddenDecl())
+    return SubclassScope::NotApplicable;
+
+  assert(FD->getEffectiveAccess() <= classType->getEffectiveAccess() &&
```

Comment:
It doesn't seem to, because `FD->getEffectiveAccess()` essentially returns `min(FD->getFormalAccess(), classType->getFormalAccess())`.

Also, FWIW, this function, including this assert and the overridden decl check was moved wholesale from SILFunction.cpp.